File: domains/org/src/Entities/Role/Role.cs
using Base.Entities;

namespace Org.Entities;

public class Role : EntityBase
{
  public int? RoleTypeId { get; set; }
  public RoleType? RoleType { get; set; }

  public string CareerTrack { get; set; } = string.Empty;

  public virtual List<RoleAssignment> RoleAssignments { get; set; } = new();
  public virtual List<LevelAssignment> LevelAssignments { get; set; } = new();
  public virtual List<ManagerAssignment> ManagerAssignments { get; set; } = new();

  public virtual List<Capability> Capabilities { get; set; } = new();

  public DateTimeOffset OpenFromDate { get; set; } = DateTimeOffset.UtcNow;
  public DateTimeOffset? ClosedAtDate { get; set; }
}

File: webapp/src/Client/Role.ts
import { useCRUD, UseCRUDOptionalProps } from "./CRUD/CRUD";
import { RoleType } from "./RoleType";
import { RoleAssignment } from "./RoleAssignment";
import { ManagerAssignment } from "./ManagerAssignment";
import { Title } from "./Title";
import { Capability } from "./Capability";

export interface TitleAssignment {
  id: number;
  title: Title;
  role: Role;
}

export interface Role {
  id: number;

  role_type?: RoleType;

  role_assignments?: RoleAssignment[];
  manager_assignments?: ManagerAssignment[];
  title_assignments: TitleAssignment[];

  start_date: string;
  end_date?: string;

  active_role_assignment?: RoleAssignment;
  active_manager_assignment?: ManagerAssignment;
  active_title_assignment: TitleAssignment;

  capabilities: Capability[];
}

export interface MutateRole {
  role_type_id: number;
  title_id: number;
  start_date?: string;
  end_date?: string;
}

export interface UseRoleProps extends UseCRUDOptionalProps {
  parentUnitId?: number;
}

export function useRole(props: UseRoleProps = {}) {
  const path = `/api/org/roles`;
  const {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  } = useCRUD<Role, MutateRole>({
    path,
    ...props,
  });
  return {
    items,
    loading,
    error,
    reload,
    createItem,
    retrieveItem,
    updateItem,
    deleteItem,
  };
}
